### ISSUE TITLE
new check: UnnecessarySemicolonInEnumeration

### ISSUE THREAD

**Issue by strkkk** [checkstyle/checkstyle#6773]
From https://github.com/checkstyle/checkstyle/issues/6752#issuecomment-493788510
When enum has nothing but constants declaration, trailing semicolon is unnecessary.
It should be detected and reported as violation

```
$ cat TestClass.java
enum Nothing {
  A,B
}
enum Comma {
  A,B,
}
enum Semicolon {
  A,B; // violation
}
enum EvenThis {
  A,B,; // violation
}

enum Normal {
A,B; // OK
Normal(){}
}

$ cat conf.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
       <module name="UnnecessarySemicolonInEnumeration" />
    </module>
</module>

$ java -jar checkstyle-8.20-all.jar -c conf.xml Test.java
Starting audit...
here is violations at lines commented with // violation
```

**Comment by romani** [comment]
please add to test case:
```
enum EvenThis {
  A,
  B,
  ; // required ";", no violation
  Normal(){}
}
```

If there is something in Enum other than elements, ";" is requirement, no violations in any form of code(minimization of diff in multiline mode is not applicable here). 

I am ok with design.

@rnveach and/or @pbludov , please review and approve.

**Comment by rnveach** [comment]
I am ok.

**Comment by pbludov** [comment]
I think we also need a check like [ArrayTrailingComma](https://checkstyle.org/config_coding.html#ArrayTrailingComma):
Every constant in _simple_ enum should ends with a comma, for the same reason.
This check will forbid semicolon at the end, and a separate `EnumConstantTrailingComma` check will require a comma at the end.

**Comment by strkkk** [comment]
@romani
> please add to test case:

I added to test cases many things (like parentheses/comma/block and its combinations) including semicolon on next line.
I included your example into docs.

Please check #6782

**Comment by romani** [comment]
> I think we also need a check like 

this is separate discussion, if you see that something new should be done, please open new issue (discussion thread).
Please confirm that you are ok with proposed design of Check in this issue.

**Comment by rnveach** [comment]
Fix was merged

**Comment by romani** [comment]
@strkkk ,


We got CI failure in master - https://circleci.com/gh/checkstyle/checkstyle/22189 :
```
Actual:
JavadocPropertiesGenerator.java.html:<td class='covered'><pre><span  class='survived'>            if (result) {</span></pre></td></tr>
Ignore:

Diff:
--- target/ignored.txt	2019-06-01 12:55:07.402379061 +0000
+++ target/actual.txt	2019-06-01 12:55:07.402379061 +0000
@@ -0,0 +1 @@
+JavadocPropertiesGenerator.java.html:<td class='covered'><pre><span  class='survived'>            if (result) {</span></pre></td></tr>
Difference between 'Actual' and 'Ignore' lists is detected, lists should be equal.
build will be failed.
Exited with code 1
```
I am not sure how this leak happened, but we need to fix it, as all further PRs will be failing in the same way.

**Comment by strkkk** [comment]
@romani this is really strange, I didn't touch this class in my PR.
I can check it later today.

**Comment by strkkk** [comment]
I played with this condition and tests always result in failure for mutations I made (remove condition, negate it)
Also, other PRs like https://github.com/checkstyle/checkstyle/pull/6797 pass pitest ok.

**Comment by rnveach** [comment]
CircleCI is passing in master now so I would assume this is a random error and not related to this issue.
It could be close to the edge of timing out, which would falsely flag it as being killed. Or it could be something random that causes to pass and fail on different machines.
